_studio/shared/src/libmfx_allocator_vaapi.cpp

```diff
@@ -110,6 +110,129 @@ unsigned int ConvertMfxFourccToVAFormat(mfxU32 fourcc)
     }
 }
 
+static mfxStatus ReallocImpl(mfxDefaultAllocatorVAAPI::mfxWideHWFrameAllocator* pSelf, vaapiMemIdInt *vaapi_mid, mfxFrameSurface1 *surf)
+{
+    VAStatus  va_res  = VA_STATUS_SUCCESS;
+    unsigned int va_fourcc = 0;
+
+    MFX_CHECK_NULL_PTR3(pSelf, vaapi_mid, surf);
+
+    mfxU32 fourcc = surf->Info.FourCC;
+
+    // VP8 hybrid driver has weird requirements for allocation of surfaces/buffers for VP8 encoding
+    // to comply with them additional logic is required to support regular and VP8 hybrid allocation pathes
+    mfxU32 mfx_fourcc = ConvertVP8FourccToMfxFourcc(fourcc);
+    va_fourcc = ConvertMfxFourccToVAFormat(mfx_fourcc);
+    if (!va_fourcc || ((VA_FOURCC_NV12 != va_fourcc) &&
+                       (VA_FOURCC_YV12 != va_fourcc) &&
+                       (VA_FOURCC_YUY2 != va_fourcc) &&
+                       (VA_FOURCC_ARGB != va_fourcc) &&
+                       (VA_FOURCC_ABGR != va_fourcc) &&
+                       (VA_FOURCC_P208 != va_fourcc) &&
+                       (VA_FOURCC_P010 != va_fourcc) &&
+                       (VA_FOURCC_YUY2 != va_fourcc) &&
+#if (MFX_VERSION >= 1027)
+                       (VA_FOURCC_Y210 != va_fourcc) &&
+                       (VA_FOURCC_Y410 != va_fourcc) &&
+#endif
+                       (VA_FOURCC_AYUV != va_fourcc) ))
+    {
+        return MFX_ERR_UNSUPPORTED;
+    }
+
+    VASurfaceID surfaces[1];
+    VASurfaceAttrib attrib[2];
+    unsigned int format;
+    int attrCnt = 0;
+
+    surfaces[0] = *vaapi_mid->m_surface;
+    if (MFX_FOURCC_P8 == vaapi_mid->m_fourcc)
+    {
+        va_res = vaDestroyBuffer(pSelf->pVADisplay, surfaces[0]);
+        if (VA_STATUS_SUCCESS != va_res)
+            return MFX_ERR_MEMORY_ALLOC;
+    }
+    va_res = vaDestroySurfaces(pSelf->pVADisplay, &surfaces[0], 1);
+    if (VA_STATUS_SUCCESS != va_res)
+        return MFX_ERR_MEMORY_ALLOC;
+
+    *vaapi_mid->m_surface = 0;
+
+    attrib[attrCnt].type          = VASurfaceAttribPixelFormat;
+    attrib[attrCnt].flags         = VA_SURFACE_ATTRIB_SETTABLE;
+    attrib[attrCnt].value.type    = VAGenericValueTypeInteger;
+    attrib[attrCnt++].value.value.i = va_fourcc;
+    format               = va_fourcc;
+
+    if (fourcc == MFX_FOURCC_VP8_NV12)
+    {
+        // special configuration for NV12 surf allocation for VP8 hybrid encoder is required
+        attrib[attrCnt].type          = (VASurfaceAttribType)VASurfaceAttribUsageHint;
+        attrib[attrCnt].flags           = VA_SURFACE_ATTRIB_SETTABLE;
+        attrib[attrCnt].value.type      = VAGenericValueTypeInteger;
+        attrib[attrCnt++].value.value.i = VA_SURFACE_ATTRIB_USAGE_HINT_ENCODER;
```

Comment:
so this code utilizes side-effects of postfix operator++ and uses dummy element at the end...

And you still suppose that this is good solution?
I find it hard to understand and very error-prone

I understand that this is legacy approach, but at some point we need to start changing code to be better. I don't see a reason why can't we start now